protocol/nats_jetstream/v3/sender.go

```diff
@@ -0,0 +1,139 @@
+/*
+ Copyright 2024 The CloudEvents Authors
+ SPDX-License-Identifier: Apache-2.0
+*/
+
+package nats_jetstream
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+
+	"github.com/nats-io/nats.go"
+	"github.com/nats-io/nats.go/jetstream"
+
+	"github.com/cloudevents/sdk-go/v2/binding"
+	"github.com/cloudevents/sdk-go/v2/protocol"
+)
+
+type sender struct {
+	// nats connection
+	conn      *nats.Conn
+	url       string
+	natsOpts  []nats.Option
+	connOwned bool
+
+	// jetstream options and configuration
+	jetSteamOpts []jetstream.JetStreamOpt
+	jetStream    jetstream.JetStream
+	publishOpts  []jetstream.PublishOpt
+
+	subject string
+}
+
+// NewSender creates a new protocol.Sender responsible for opening and closing the NATS connection
+// Use WithURL() OR WithConnection()
+func NewSender(ctx context.Context, subject string, opts ...ProtocolOption) (*sender, error) {
+	s := &sender{
+		subject: subject,
+	}
+
+	if err := s.applyOptions(opts...); err != nil {
+		return nil, err
+	}
+
+	if err := s.validateOptions(); err != nil {
+		return nil, err
+	}
+
+	var errConnection error
+	defer func() {
+		// close connection if an error occurs and we created the nats connection
+		if s.connOwned && s.conn != nil && errConnection != nil {
+			s.conn.Close()
+		}
+	}()
+
+	// create a connection based on the URL if the connection was not given as an option
+	if s.conn == nil {
+		s.conn, errConnection = nats.Connect(s.url, s.natsOpts...)
+		if errConnection != nil {
+			return nil, errConnection
+		}
+	}
+
+	var err error
+	if s.jetStream, err = jetstream.New(s.conn, s.jetSteamOpts...); err != nil {
+		return nil, err
+	}
+
+	return s, nil
+}
+
+// Close implements Sender.Sender
+// Sender sends messages.
+func (s *sender) Send(ctx context.Context, in binding.Message, transformers ...binding.Transformer) (err error) {
+	defer func() {
+		if err2 := in.Finish(err); err2 != nil {
+			if err == nil {
+				err = err2
+			} else {
+				err = fmt.Errorf("failed to call in.Finish() when error already occurred: %s: %w", err2.Error(), err)
+			}
+		}
+	}()
+
+	if _, err = s.jetStream.StreamNameBySubject(ctx, s.subject); err != nil {
+		return err
+	}
+
+	writer := new(bytes.Buffer)
+	header, err := WriteMsg(ctx, in, writer, transformers...)
+	if err != nil {
+		return err
+	}
+
+	natsMsg := &nats.Msg{
+		Subject: s.subject,
+		Data:    writer.Bytes(),
+		Header:  header,
+	}
+
+	_, err = s.jetStream.PublishMsg(ctx, natsMsg, s.publishOpts...)
+
+	return err
+}
+
+// Close implements Closer.Close
+// This method only closes the connection if the Sender opened it
+func (s *sender) Close(_ context.Context) error {
```

Comment:
Question: do we need those Close() methods on the sender instead of at the protocol layer?